### ISSUE TITLE
#units is wrong for MNIST 4x1024

### ISSUE THREAD

**Issue by stanleybak** [eth-sri/eran#14]
In the table in the `README` (and I believe in the papers as well if this is meant to be `FFNNBig`), it says the network `MNIST`, `4x1024`, `fully connected` (filename: `mnist_relu_4_1024.tf`) has `4106` hidden units.

This network takes in a `784` element vector (`28x28` image), has 3 ReLU layers that produce a `1024` element output, followed by an Affine layer that produces a `10` element output. By my count, that's `1024*3 + 10 = 3082` hidden units. Am I missing something?

**Comment by stanleybak** [comment]
The name `FFNNBig` also seems to refer to different networks depending on the paper. In "Fast and effective robustness certification" FFNNBig  has 3010 hidden units, whereas in "An abstract domain for verifying neural networks" FFNNBig has 4106 hidden units. I guess 4x1024 is the one in "An abstract domain for verifying neural networks", although it would be better if this was more clear.

**Comment by GgnDpSngh** [comment]
Hi there,

Thanks for your interest in ERAN. The 4x1024 network has 3 hidden layers with ReLU activation followed by an affine layer with no activation. Each of these have 1024 neurons so the total number of hidden neurons is 4106. Note that this network is the same as 4x1024 network used in Fast-lin paper (https://arxiv.org/pdf/1804.09699.pdf).

You are right, FFNNBig refers to different networks in the NeurIPS'18 and POPL'19 papers and the 4x1024 network refers to FFNNBig in the POPL'19 paper.

Let me know if you have any other queries.

Cheers,

**Comment by stanleybak** [comment]
Thanks for the quick reply and extra info. 

What are you counting as a neuron in this case? Shouldn't the number of neurons in a layer equal the number of outputs of the layer? In this case, the affine layer has 10 outputs (not 1024).

**Comment by GgnDpSngh** [comment]
Hi there,

We investigated the network further and you are indeed right, the network contains 3,082 neurons, we will update the site and paper accordingly.

Thanks for pointing this out.

Cheers,

**Comment by stanleybak** [comment]
I believe some of the other networks also have problems with the reported number of hidden units. The first one in the readme table, MNIST 3x50 is written as having 160 units, but the network in the `mnist_relu_3_50.tf` file has 3 ReLU layers with 50, 50, and 10 outputs, so that seems to me like it should be 110 hidden units.

**Comment by GgnDpSngh** [comment]
Hi there,

You are right regarding the neuron counting issue for the 3x50 network as well. We have noticed that for a few other feedforward networks, the reported number for hidden neurons is one layer more than the actual number. We also informed about this bug to the FastLin authors that trained the 4x1024 network and they confirmed it.  

We have now added new functionality in ERAN that also reports the total number of hidden neurons in the network. We will update the site and paper with the correct numbers shortly.

Cheers,

**Comment by GgnDpSngh** [comment]
Hi there,

We have updated the website with the correct number of neurons for each network. Here #neurons=number of activation units (ReLU, Sigmoid, Tanh etc). While the neuron count for feedforward networks is now less than before, the neuron count for few convolutional networks reported earlier was less than actual. For example, the MNIST ConvBig network has around 48K neuron while the earlier count was 35K. Let me know if you find any other issue on the neuron counting.

Cheers,
